File: src/features/ai_engine/forecasting_engine.ts
export class ForecastingEngine {
  /**
   * Simple Linear Regression to predict next value
   * @param y List of values (e.g., monthly spending)
   */
  static predictNext(y: number[]): number {
    if (y.length < 2) return y.length > 0 ? y[y.length - 1] : 0.0;

    const n = y.length;
    const x = Array.from({ length: n }, (_, i) => i);

    const sumX = x.reduce((a, b) => a + b, 0);
    const sumY = y.reduce((a, b) => a + b, 0);
    let sumXY = 0;
    let sumXX = 0;

    for (let i = 0; i < n; i++) {
      sumXY += x[i] * y[i];
      sumXX += x[i] * x[i];
    }

    const slope = (n * sumXY - sumX * sumY) / (n * sumXX - sumX * sumX);
    const intercept = (sumY - slope * sumX) / n;

    // Predict for the next index (n)
    return slope * n + intercept;
  }

  static movingAverage(data: number[], period: number = 3): number {
    if (data.length === 0) return 0.0;
    const actualPeriod = data.length < period ? data.length : period;
    const lastN = data.slice(data.length - actualPeriod);
    return lastN.reduce((a, b) => a + b, 0) / actualPeriod;
  }
}

File: flutter_backup/lib/features/ai_engine/forecasting/forecasting_engine.dart
class ForecastingEngine {
  /// Simple Linear Regression to predict next value
  /// [y] is the list of values (e.g., monthly spending)
  static double predictNext(List<double> y) {
    if (y.length < 2) return y.isNotEmpty ? y.last : 0.0;

    int n = y.length;
    List<int> x = List.generate(n, (i) => i);

    double sumX = x.reduce((a, b) => a + b).toDouble();
    double sumY = y.reduce((a, b) => a + b);
    double sumXY = 0;
    double sumXX = 0;

    for (int i = 0; i < n; i++) {
      sumXY += x[i] * y[i];
      sumXX += x[i] * x[i];
    }

    double slope = (n * sumXY - sumX * sumY) / (n * sumXX - sumX * sumX);
    double intercept = (sumY - slope * sumX) / n;

    // Predict for the next index (n)
    return slope * n + intercept;
  }

  static double movingAverage(List<double> data, {int period = 3}) {
    if (data.isEmpty) return 0.0;
    int actualPeriod = data.length < period ? data.length : period;
    List<double> lastN = data.sublist(data.length - actualPeriod);
    return lastN.reduce((a, b) => a + b) / actualPeriod;
  }
}
